src/java/org/apache/cassandra/index/sai/plan/OrderByProcessor.java

```diff
@@ -0,0 +1,214 @@
+ * - for the first query, coordinator returns global top page-size rows within entire ring
+ * - for the subsequent queries, coordinators returns global top page-size rows withom range from last-returned-row to max token
+ */
+public class OrderByProcessor
+{
+    private final ReadCommand command;
+    private final StorageAttachedIndexQueryPlan queryPlan;
+
+    private final int limit;
+    private final Set<PrimaryKey> tombstonesSeen;
+    private final IndexContext indexContext;
+
+    public OrderByProcessor(ReadCommand command, StorageAttachedIndexQueryPlan queryPlan, IndexContext indexContext)
+    {
+        this.command = command;
+        this.queryPlan = queryPlan;
+        this.limit = command.limits().count();
+        this.indexContext = indexContext;
+        this.tombstonesSeen = new HashSet<>();
+    }
+
+    /**
+     * Filter given partitions and keep the rows with highest scores. In case of {@link UnfilteredPartitionIterator},
+     * all tombstones will be kept.
+     */
+    public <U extends Unfiltered, R extends BaseRowIterator<U>, P extends BasePartitionIterator<R>> BasePartitionIterator<?> filter(P partitions)
+    {
+        // priority queue ordered by score in ascending order
+        PriorityQueue<Triple<PartitionInfo, Row, Float>> topK = new PriorityQueue<>(limit, Comparator.comparing(Triple::getRight));
+        // to store top-k results in primary key order
+        TreeMap<PartitionInfo, TreeSet<Unfiltered>> unfilteredByPartition = new TreeMap<>(Comparator.comparing(p -> p.key));
+
+        Map<StorageAttachedIndex, float[]> queryVectorPerIndex = getQueryVectorPerIndex(command);
+        while (partitions.hasNext())
+        {
+            try (R partition = partitions.next())
+            {
+                DecoratedKey key = partition.partitionKey();
+                Row staticRow = partition.staticRow();
+                PartitionInfo partitionInfo = PartitionInfo.create(partition);
+                // compute key and static row score once per partition
+                float keyAndStaticScore = getScoreForRow(queryVectorPerIndex, key, staticRow);
+
+                while (partition.hasNext())
+                {
+                    Unfiltered unfiltered = partition.next();
+                    // Always include tombstones for coordinator. It relies on ReadCommand#withMetricsRecording to throw
+                    // TombstoneOverwhelmingException to prevent OOM.
+                    if (!unfiltered.isRow())
+                    {
+                        unfilteredByPartition.computeIfAbsent(partitionInfo, k -> new TreeSet<>(command.metadata().comparator))
+                                             .add(unfiltered);
+                        continue;
+                    }
+
+                    Row row = (Row) unfiltered;
+                    if (!row.hasLiveData(command.nowInSec(), true))
+                        tombstonesSeen.add(indexContext.keyFactory().create(key, row.clustering()));
```

Comment:
this only includes row tombstones, how about partition deletion and range tombstones?

Reply:
I'm hopeful that if we hook similar logic up in the correct place, which this is not, then it will be easier to handle those as well :)